cmd/containerd-shim-runhcs-v1/task_hcs.go

```diff
@@ -1008,3 +1026,75 @@ func (ht *hcsTask) requestUpdateContainer(ctx context.Context, resourcePath stri
 	}
 	return ht.c.Modify(ctx, modification)
 }
+
+// publishNotifications monitors the containers notification queue.
+// Currently only OOM events are published.
+//
+// This MUST be called via a goroutine to wait on a background thread.
+// Returned error is used for scoping, and not meant to be consumed
+func (ht *hcsTask) publishNotifications(ctx context.Context) {
+	var (
+		err error
+		nq  *queue.MessageQueue
+	)
+	ctx, span := trace.StartSpan(ctx, "hcsTask::publishNotifications")
+	defer span.End()
+	defer func() { oc.SetSpanStatus(span, err) }()
+	span.AddAttributes(trace.StringAttribute(logfields.TaskID, ht.id))
+	entity := log.G(ctx).WithField(logfields.TaskID, ht.id)
+
+	nq, err = ht.c.Notifications()
+	if err != nil {
+		// includes `notifications.ErrNotificationsNotSupported`, where container
+		// does not publish any notifications
+		entity.WithError(err).Warning("unable to retrieve container notifications channel")
+		return
+	}
+
+	for {
+		select {
+		case <-ctx.Done():
+			return
+		default:
+			var v interface{}
+			v, err = nq.ReadOrWait()
```

Comment:
Might be ideal here to be waiting for one of two cases to fire and have no default. Either we get a notification and some `notifChan` fires and then we handle the event, or the context is done and we go straight to that case. With what's here currently we'll more than likely end up in the default case a majority of the time which means we'll end up blocking until `ReadOrWait` blesses us with a value. So even if the context is closed or cancelled we'd still be blocked on `ReadOrWait` for an indefinite amount of time. Goroutines are fairly cheap so I don't think it's too much of a stretch to do something like the following:

```go
notifChan := make(chan interface{})
go func(){
    for {
        v, err = nq.ReadOrWait()
        // handle error
        notifChan <- v
    }
}()


for {
    select {
         case <- ctx.Done():
             return
         case notif := <-notifChan:
            // use notif
    }
}

Reply:
Reasoning is we don't really expect a whole bunch of notifications to be occurring, so ReadOrWait will likely be blocked for most of its life. Really just OOM (which we hope doesn't happen) and when a container exits iirc right?

Reply:
`ReadOrWait` will return if the notifications channel is closed, which happens if close is called on the container as well as if it exits in the guest.
So I guess if the context passed to `newHcsTask` was cancelled, but close wasn't called on the task or container, it would hang indefinitely.

Is that a scenario we should look at fixing? 
As in, should `gcs.Container` (and co) monitor for `ctx.Done` in `waitBackground`?

Reply:
Ehh, in that case this is fine. Didn't realize the notif channel is closed on container close